src/zenml/models/hub_plugin_models.py

```diff
@@ -0,0 +1,76 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Models representing ZenML Hub plugins."""
+
+from datetime import datetime
+from typing import TYPE_CHECKING, List, Optional
+from uuid import UUID
+
+from pydantic import BaseModel
+
+from zenml.utils.enum_utils import StrEnum
+
+if TYPE_CHECKING:
+    pass
+
+
+class PluginStatus(StrEnum):
+    """Enum that represents the status of a plugin."""
+
+    PENDING = "pending"  # Plugin is being built
+    FAILED = "failed"  # Plugin build failed
+    AVAILABLE = "available"  # Plugin is available for installation
+    YANKED = "yanked"  # Plugin was yanked and is no longer available
+
+
+class HubUserResponseModel(BaseModel):
+    """Model for a ZenML Hub user."""
+
+    id: UUID
+    email: str
+    username: Optional[str]
+
+
+class HubPluginBaseModel(BaseModel):
+    """Base model for a ZenML Hub plugin."""
+
+    name: str
+    description: Optional[str]
+    version: Optional[str]
+    release_notes: Optional[str]
+    repository_url: str
+    repository_subdirectory: Optional[str]
+    repository_branch: Optional[str]
```

Comment:
With the recent mypy changes, I think these should all be `Optional[...] = None`.

Also: can both `commit` and `branch` be set at the same time? What does branch even mean in this case, I would expect it to build a new plugin if the branch gets updated.

Reply:
yeah, already changed the types.

As for `commit` and `branch`, `branch` can only be set at creation time will be ignored if `commit` is set. When the plugin is being built, we also override `commit` according to the actual repo commit that we built the plugin from, so `commit` will always be set for plugins that were built correctly, but might be missing for plugins where the build failed or is still in progress.

Reply:
Wouldn't it be cleaner to have the `branch` only on the request model in that case?

Reply:
Client-side yes, but the models that we have here are identical to the models we have server-side and I would like to keep them in sync to avoid unnecessary complexity.

Server-side we need the branch on the response model as well because it currently works like this:
- client submits plugin -> validate plugin -> write plugin to DB with status "pending" -> return corresponding response model
- plugin build is triggered with response model as input -> plugin is built -> plugin is updated in DB with status "failed" or "available" (and only now the branch is not needed anymore)